### PR TITLE
Use AVX instructions to compute IBD faster

### PR DESCRIPTION
**Pull request by danking**
This PR implements the core IBS operations in terms of vectorized C code. In particular, we use the `libsimdpp` library to take advantage of whatever the widest available register is (many modern CPUs have AVX2 256 bit integer registers; Knights Landing will introduce AVX512 512-bit integer registers).

The performance improvement is massive. We can compute the full IBD matrix on 2,535 samples and ~37 million variants in just under 17 minutes. We believe the complexity of this code is `O(nSamples^2 * nVariants)`. Assuming the scaling works out well, we should be able to compute 100,000 Variants and 40,000 samples in the same time.

There were a couple issues I had to workaround, but hopefully we can re-use those workarounds:

- compiling native code from gradle
- packaging native code for `test`, `installDist`, and the JARs
- building native code specialized to certain architectures

Still left to do:

- [x] break the C tests into a separate file and call from gradle `test`
- [x] maybe use a library ([libsimdpp?](https://github.com/p12tic/libsimdpp)) to do the SIMD so we're agnostic to the underlying architecture (right now if you don't have AVX, we fall all the way back to 64-bit registers, rather than 128-bit SSE registers) 
- [x] some minor clean up of the IBSFFI class

Future Work:
- implement IBSExpectations in C as well
- expand this work to KING (or other structure correcting IBD calculations)

### CODE REVIEW

**Comment by danking** [comment]
We're 11x slower than plink now:

```bash
# time plink --bfile profile225 --genome 
PLINK v1.90b3.38 64-bit (7 Jun 2016)       https://www.cog-genomics.org/plink2
(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to plink.log.
Options in effect:
  --bfile profile225
  --genome

16384 MB RAM detected; reserving 8192 MB for main workspace.
224885 variants loaded from .bim file.
2535 people (0 males, 0 females, 2535 ambiguous) loaded from .fam.
Ambiguous sex IDs written to plink.nosex .
Using up to 4 threads (change this with --threads).
Before main variant filters, 2535 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate is 0.952416.
224885 variants and 2535 people pass filters and QC.
Note: No phenotypes present.
IBD calculations complete.  
Finished writing plink.genome .
plink --bfile profile225 --genome  67.81s user 1.03s system 297% cpu 23.147 total

# time ../hail/build/install/hail/bin/hail read -i profile225-splitmulti.vds ibd -o hail.genome
hail: info: running: read -i profile225-splitmulti.vds
[Stage 0:>                                                          (0 + 0) / 4]SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[Stage 1:============================================>              (3 + 1) / 4]hail: info: running: ibd -o hail.genome
[Stage 8:======================================================>  (62 + 3) / 65]hail: info: while writing:
    hail.genome
  merge time: 6.980s
hail: info: timing:
  read: 2.953s
  ibd: 4m12.3s
  total: 4m15.2s
../hail/build/install/hail/bin/hail read -i profile225-splitmulti.vds ibd -o   840.77s user 23.05s system 332% cpu 4:19.75 total

# dc
60 4 * 19 +
5 k
23 / p
11.26086
```

**Comment by danking** [comment]
2535 Samples, 236,734 Variants. We expect this computation to scale quadratically in sample size.

**Comment by danking** [comment]
Since everyone is asking about hardcalls:

```
# (cd ../hail && gradle installDist) && ../hail/build/install/hail/bin/hail read -i profile225-splitmulti-hardcalls.vds ibd -o hail.genome                 
:nativeLib UP-TO-DATE
:compileJava UP-TO-DATE
:compileScala UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar UP-TO-DATE
:startScripts UP-TO-DATE
:installDist UP-TO-DATE

BUILD SUCCESSFUL

Total time: 2.728 secs
hail: info: running: read -i profile225-splitmulti-hardcalls.vds
[Stage 0:>                                                          (0 + 0) / 4]SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[Stage 1:============================================>              (3 + 1) / 4]hail: info: running: ibd -o hail.genome
[Stage 8:======================================================>  (62 + 3) / 65]hail: info: while writing:
    hail.genome
  merge time: 6.619s
hail: info: timing:
  read: 3.824s
  ibd: 3m19.2s
  total: 3m23.1s

# dc
5 k
3 60 * 23 + 
23 / p
8.82608
```

about 9x slower now.

**Comment by danking** [comment]
In both cases we had 279 cores, 1:1 core:partition

With 37 million variants

```
nSamples             2,535
nVariants       37,237,504
```
```
dking@cluster-2-m:~/hail$ spark-submit   --class org.broadinstitute.hail.driver.Main   build/libs/hail-all-spark.jar read -i gs://hail-1kg/ALL.1KG.qc.hardcalls.vds/ ibd -o gs://hail-1kg/ALL.1KG.qc.genome --parallel-write
hail: info: running: read -i gs://hail-1kg/ALL.1KG.qc.hardcalls.vds/
[Stage 1:=================================================>         (5 + 1) / 6]hail: info: running: ibd -o gs://hail-1kg/ALL.1KG.qc.genome --parallel-write
[Stage 8:======================================================>(278 + 1) / 279]hail: info: timing:
  read: 27.197s
  ibd: 16m21.3s
  total: 16m48.5s
```

With 76 thousand variants

```
nSamples             2,535
nVariants           76,628
```
```
dking@cluster-2-m:~/hail$ spark-submit   --class org.broadinstitute.hail.driver.Main   build/libs/hail-all-spark.jar read -i gs://hail-1kg/ALL.1KG.qc.hardcalls.filtered.AF005.callrate98.p1000.vds/ filtervariants expr --keep -c 'v.start % 200 == 0' ibd -o gs://hail-1kg/ALL.1KG.callrate98.p1000.qc.genome
hail: info: running: read -i gs://hail-1kg/ALL.1KG.qc.hardcalls.filtered.AF005.callrate98.p1000.vds/
[Stage 1:======================================================>(269 + 1) / 270]hail: info: running: filtervariants expr --keep -c 'v.start % 200 == 0'
hail: info: running: ibd -o gs://hail-1kg/ALL.1KG.callrate98.p1000.qc.genome
[Stage 8:======================================================>(278 + 1) / 279]hail: info: while writing:
    gs://hail-1kg/ALL.1KG.callrate98.p1000.qc.genome
  merge time: 3.590s
hail: info: timing:
  read: 40.466s
  filtervariants expr: 288.900ms
  ibd: 40.869s
  total: 1m21.6s
```

**Comment by danking** [comment]
I'm done, this is ready for review 🎉

**Comment by cseed** [comment]
I tried to look at this, but libsimdpp is completely spamming the diff visualizer.  I'm back to thinking we shouldn't include this in the repo.  We should either assume it is installed or download it during the build process.  I'm inclined to do the former for now.

**Comment by danking** [comment]
👍 We can use the [tar](https://github.com/p12tic/libsimdpp/archive/v2.0-rc2.tar.gz) that he published [on github](https://github.com/p12tic/libsimdpp/releases/tag/v2.0-rc2). I'll get that change in today.

**Comment by danking** [comment]
@cseed back to you.

File: src/main/c/ibs.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+using namespace simdpp;
+
+#ifndef HAIL_OVERRIDE_WIDTH
+#define HAIL_VECTOR_WIDTH SIMDPP_FAST_INT64_SIZE
+#else
+#define HAIL_VECTOR_WIDTH HAIL_OVERRIDE_WIDTH
+#endif // HAIL_OVERRIDE_WIDTH
+
+using hailvec = uint64<HAIL_VECTOR_WIDTH>;
```
Comment (danking) [inline src/main/c/ibs.h]: Changed.

File: src/main/c/ibs.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+#ifndef CACHE_SIZE_PER_MATRIX_IN_KB
+#define CACHE_SIZE_PER_MATRIX_IN_KB 4
+#endif
+
+#ifndef CACHE_SIZE_IN_MATRIX_ROWS
+#define CACHE_SIZE_IN_MATRIX_ROWS 4 * CACHE_SIZE_PER_MATRIX_IN_KB
```
Comment (danking) [inline src/main/c/ibs.h]: I cleaned up the code a fair bit, but then made this bit more complicated. I think, however, that it's more straightforward now? Let me know what you think.

File: src/main/c/ibs.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            uint64_t* __restrict__ y,
+            hailvec * __restrict__ x_na_masks,
+            hailvec * __restrict__ y_na_masks) {
+  uint64_t i = 0;
+  for (; i <= (length - HAIL_VECTOR_WIDTH); i += HAIL_VECTOR_WIDTH) {
+    hailvec x256 = load_u(x+i);
+    hailvec y256 = load_u(y+i);
+    ibs256(result, x256, y256, x_na_masks[i/HAIL_VECTOR_WIDTH], y_na_masks[i/HAIL_VECTOR_WIDTH]);
+  }
+  for (; i < length; ++i) {
```
Comment (danking) [inline src/main/c/ibs.cpp]: But am now.

**Comment by danking** [comment]
@cseed back to you

**Comment by cseed** [comment]
Sorry I took so long on this.  It's good to go once you resolve the conflicts.

**Comment by cseed** [comment]
I resolved them with Github's resolution tool (nice!)  Will merge once it passes CI.

**Pull request merged**